nbgrader/docs/source/user_guide/what_is_nbgrader.rst

```diff
@@ -0,0 +1,308 @@
+What is nbgrader?
+=================
+
+This page describes the internals of nbgrader from a rather technical
+standpoint.  It is needed not just for developers, but also for
+administrators which need to build advanced deployments, possibly
+customizing various components in the process.
+
+
+
+Nbgrader format
+---------------
+
+Nbgrader uses the standard Jupyter notebook format, ``.ipynb``.  This
+is the primary power of nbgrader: since it uses a standard format
+that's used for other types of computation.  Since it is not special
+to nbgrader, it is immediately usable on other systems (running on
+your own computer, uploading to Google Colab, using anything from the
+Jupyter ecosystem).  Thus, to begin developing nbgrader notebooks,
+nothing special is needed.  They are directly portable to research
+code, other projects, etc.  The skills learned in doing nbgrader
+assignments is portable to other real-world projects.
+
+All of the files are stored directly on a filesystem, so that they are
+easy to manipulate and understand.  The text (well, JSON) based
+notebook format is well supported by Jupyter tools for manipulation,
+making it easy to script other tools with them.
+
+`Jupyter notebooks can have metadata
+<https://nbformat.readthedocs.io/en/latest/format_description.html>`__ -
+both notebook-wide and per-cell.  This is used to track the progress
+of the assignment.  For example, this tracks cells that are
+read-only.  However, since the metadata is within the notebook file
+itself - which is given to the student - so one can't prevent the
+student from editing the notebook file to change this.  More on the
+effects of this later.
+
+The standard notebook format is also one of the downsides, compared to
+other learning management systems (depending on if you consider this a
+disadvantage...).  Because this is not a "captive system" that
+segments users from the data they are working on, you can't easily
+limit access to the underlying operating system or notebook file.
+However, if you design your course considering this, we hope you'll
+see that this is an advantage.
+
+
```

Comment:
Working with Nbgrader
------------------------------

When nbgrader was first written, all interactions were done on the command line.

Later on it acquired notebook extensions to enable interaction from within a jupyter notebook.

These notes will refer to the command-line interface, and we'll discuss the UI interactions later.